trin-execution/src/subcommands/era2.rs

```diff
@@ -0,0 +1,243 @@
+                    });
+                }
+            }
+
+            // Get the rounded up storage count
+            let storage_count = storage.len().div_ceil(MAX_STORAGE_ITEMS);
+
+            era2.append_entry(&AccountOrStorageEntry::Account(AccountEntry {
+                address_hash: account_hash,
+                account_state: (&account_state).into(),
+                bytecode,
+                storage_count: storage_count as u32,
+            }))?;
+
+            for storage_chunk in storage.chunks(MAX_STORAGE_ITEMS) {
+                era2.append_entry(&AccountOrStorageEntry::Storage(StorageEntry(
+                    storage_chunk.to_vec(),
+                )))?;
+            }
+
+            accounts_processed += 1;
+            if accounts_processed % 10000 == 0 {
+                info!("Processed {} accounts", accounts_processed);
+            }
+        }
+
+        info!("Era2 snapshot exported");
+
+        Ok(())
+    }
+}
+
+pub struct StateImporter {
+    pub trin_execution: TrinExecution,
+    importer_config: ImportStateConfig,
+}
+
+impl StateImporter {
+    pub fn new(trin_execution: TrinExecution, importer_config: ImportStateConfig) -> Self {
+        Self {
+            trin_execution,
+            importer_config,
+        }
+    }
+
+    pub fn import_state(&mut self) -> anyhow::Result<()> {
+        info!("Importing state from .era2 file");
+        if self.trin_execution.next_block_number() != 0 {
+            return Err(Error::msg(
+                "Cannot import state from .era2, database is not empty",
+            ));
+        }
+
+        let mut era2 = Era2::open(self.importer_config.path_to_era2.clone())?;
+        info!("Era2 reader initiated");
+        let mut accounts_imported = 0;
+        while let Some(account) = era2.next() {
+            let AccountOrStorageEntry::Account(account) = account else {
+                return Err(Error::msg("Expected account, got storage entry"));
+            };
+            let AccountEntry {
+                address_hash,
+                account_state,
+                bytecode,
+                storage_count,
+            } = account;
+
+            let account_state: Account = account_state.into();
+
+            // Build storage trie
+            let account_db = AccountDB::new(address_hash, self.trin_execution.database.db.clone());
+            let mut account_trie = EthTrie::new(Arc::new(account_db));
+            for _ in 0..storage_count {
+                let Some(AccountOrStorageEntry::Storage(storage_entry)) = era2.next() else {
+                    return Err(Error::msg("Expected storage, got account entry"));
+                };
+                for StorageItem {
+                    storage_index_hash,
+                    value,
+                } in storage_entry.0
+                {
+                    account_trie
+                        .insert(storage_index_hash.as_slice(), &alloy_rlp::encode(value))?;
+                }
```

Comment:
Should we call `account_trie.root_hash()?` once per `StorageEntry`, in order to commit change to db and lower RAM usage?

Reply:
sure